src/main/java/de/unijena/cheminf/fragment/fingerprint/FragmentFingerprinter.java

```diff
@@ -418,19 +507,18 @@ public int[] getBitArray(Map<String,Integer> aUniqueSmilesToFrequencyMap) throws
     public int[] getCountArray(Map<String, Integer> aUniqueSmilesToFrequencyMap) throws NullPointerException, IllegalArgumentException {
         Objects.requireNonNull(aUniqueSmilesToFrequencyMap, "aUniqueSmilesToFrequencyMap (Map of string and integer instances) is null.");
         List<String> tmpListOfUniqueSmiles = new ArrayList<>(aUniqueSmilesToFrequencyMap.size());
-        for(String tmpUniqueSmiles : aUniqueSmilesToFrequencyMap.keySet()) {
-            if(tmpUniqueSmiles == null || aUniqueSmilesToFrequencyMap.get(tmpUniqueSmiles) == null) {
-                throw new NullPointerException("aUniqueSmilesToFrequencyMap (Map of string and integer instances) contains " +
+        for (Map.Entry<String, Integer> tmpEntry : aUniqueSmilesToFrequencyMap.entrySet()) {
+            if (tmpEntry.getKey() == null || tmpEntry.getValue() == null) {
+                throw new NullPointerException("Given map of string and integer instances contains " +
                         "instances that are null.");
+            } else if (tmpEntry.getKey().isEmpty() || tmpEntry.getKey().isBlank()) {
+                throw new IllegalArgumentException("Given map of strings an integer instances contains strings that are blank/empty.");
             }
-            if(tmpUniqueSmiles.isBlank() || tmpUniqueSmiles.isEmpty()) {
-                throw new IllegalArgumentException("aUniqueSmilesToFrequencyMap (Map of strings an integer instances) contains strings that are blank/empty.");
-            }
-            for(int i = 1; i<=aUniqueSmilesToFrequencyMap.get(tmpUniqueSmiles); i++) {
-                tmpListOfUniqueSmiles.add(tmpUniqueSmiles);
+            for(int i = 1; i <= tmpEntry.getValue(); i++) {
+                tmpListOfUniqueSmiles.add(tmpEntry.getKey());
             }
         }
-       return this.createCountArray(tmpListOfUniqueSmiles);
+        return this.createCountArray(tmpListOfUniqueSmiles);
```

Comment:
Well, this should actually be the other way around. The method with the already "more structured" input (the SMILES code to frequency map here) should internally be called by the method with a less structures input (the list of SMILES codes here). But who cares, leave it like that now.